CMakeLists.txt

```diff
@@ -0,0 +1,236 @@
+
+cmake_minimum_required(VERSION 3.15)
+cmake_policy(SET CMP0048 NEW) # project version
+cmake_policy(SET CMP0076 NEW) # full paths
+
+list(APPEND CMAKE_MODULE_PATH "${CMAKE_CURRENT_LIST_DIR}/cmake_modules")
+
+########################################################################
+# Project Details
+project(FreeRTOS-Plus-TCP
+        VERSION 3.1.0
+        DESCRIPTION "FreeRTOS TCP/UDP Network Layer"
+        HOMEPAGE_URL https://freertos.org/FreeRTOS-Plus/FreeRTOS_Plus_TCP/index.html
+        LANGUAGES C)
+
+# Do not allow in-source build.
+if( ${PROJECT_SOURCE_DIR} STREQUAL ${PROJECT_BINARY_DIR} )
+message( FATAL_ERROR "In-source build is not allowed. Please build in a separate directory, such as ${PROJECT_SOURCE_DIR}/build." )
+endif()
+
+# Options
+option(FREERTOS_PLUS_TCP_BUILD_TEST "Build the test for FreeRTOS Plus TCP" OFF)
+
+# Configuration
+# Override these at project level with:
+#   Optional: set(FREERTOS_PLUS_TCP_BUFFER_ALLOCATION "1" CACHE STRING "" FORCE)
+#   Optional: set(FREERTOS_PLUS_TCP_COMPILER "" CACHE STRING "" FORCE)
+#   Required: set(FREERTOS_PLUS_TCP_NETWORK_IF "POSIX" CACHE STRING "" FORCE)
+
+# Select the appropriate buffer allocaiton method.
+# See: https://freertos.org/FreeRTOS-Plus/FreeRTOS_Plus_TCP/Embedded_Ethernet_Buffer_Management.html
+if (NOT FREERTOS_PLUS_TCP_BUFFER_ALLOCATION)
+    message(STATUS "Using default FREERTOS_PLUS_TCP_BUFFER_ALLOCATION = 2")
+    set(FREERTOS_PLUS_TCP_BUFFER_ALLOCATION "2" CACHE STRING "FreeRTOS buffer allocation model number. 1 .. 2.")
+endif()
+
+# Select the Compiler - if left blank will detect using CMake
+# Note relies on CMake to detect over any setting here.
+# Valid options are:
+#   FREERTOS_PLUS_TCP_COMPILER | Detected | CMake
+#   -------------------------------------------------
+#   CCS                        | No       | ?TBD?
+#   GCC                        | Yes      | GNU
+#   IAR                        | Yes      | IAR
+#   Keil                       | Yes      | ARMCC
+#   MSVC                       | Yes      | MSVC # Note only for MinGW
+#   Renesas                    | No       | ?TBD?
+# Will always a attempt to detect and if detectable double checks that the compiler is set correctly.
+set(FREERTOS_PLUS_TCP_COMPILER "" CACHE STRING "FreeRTOS Plus TCP Compiler Selection")
+
+
+# Select the appropriate network interface
+# This will fail the CMake preparation step if not set to one of those values.
+set(FREERTOS_PLUS_TCP_NETWORK_IF "" CACHE STRING "FreeRTOS Plus TCP Network Interface selection")
```

Comment:
Could each of these be a separate static lib target?

Reply:
Yes - it makes more sense to separate these out as independent `CMakeLists.txt` in each of the respective portable/* directories, but I followed the way the FreeRTOS-Kernel does it instead.

That way the only selection (single case/if-else selection) is in the `portable/CMakeLists.txt` - as to which `add_subdirectory()` to use.

Reply:
Yes - independent CMakeLists.txt now - which makes adding and maintaining the ports much simpler as they are now contained with their own CMakeLists.txt file.